consul-lambda-registrator/extension/extension.go

```diff
@@ -83,26 +84,45 @@ func (ext *Extension) Start(ctx context.Context) error {
 	}()
 	defer cancel()
 
-	// Run until either the proxy returns or until the event processing loop returns.
 	errChan := make(chan error)
 
-	go ext.runProxy(ctx, errChan)
+	// The first time the proxy starts we need to wait for it to be completely
+	// initialized and ready to accept connections before we start event processing.
+	// Starting the event processing loop signals to the Lambda runtime that the extension
+	// has completed its initialization.
+	// The runProxy func will close the readyChan once the proxy is initialized and ready.
+	readyChan := make(chan struct{})
+	go ext.runProxy(ctx, readyChan, errChan)
```

Comment:
It might be better to handle this in the proxy itself since this adds up to 3 seconds to cold start times. Is it possible to make requests to the proxy and wait for the proxy's data to be retrieved if it isn't available yet?

If necessary, it's fine to do this after the beta, but we should do it before GA.

Reply:
The 3 seconds of cold start time is only if something is terribly wrong and the proxy can't start. In the unit tests I have the wait time set to 1ms.. the proxy's init only takes 20-30us but the `events.Next()` happens faster. The proxy already has the `Wait()` method which is used below.

The reason that I did it this way is the result of the existing concurrency model, which we've agreed is sub-optimal and overly complex:
- This `Start` method needs to block since it's the main extension loop.
- The `runProxy` func needs to run in a separate go routine because it has to perform the extension data refresh and proxy reconfiguration.
- The `runEvents` func needs to run in a separate go routine to handle the async events from the Lambda runtime
- On startup only `runEvents` has to wait for the proxy init to complete.

I have created a ticket to fix the concurrency model for GA so I propose addressing this then.

Please correct me if I didn't understand your question.

Reply:
Chatted with Eric about this and we decided that we are going to update the extension so that we fetch the extension data asynchronously. I will update the configuration of the dial function so that it is the one that waits on the call to fetch the extension data from parameter store. That will opimize the extension init sequence as much as we possibly can for now.